app/lib/action_view/component.rb

```diff
@@ -4,28 +4,68 @@ class Component < ActionView::Base
 
     # Entrypoint for rendering. Called by ActionView::RenderingHelper#render.
     #
-    # Returns ActionView::OutputBuffer.
-    def render_in(view_context, &block)
+    # view_context: ActionView context from calling view
+    # args(hash): params to be passed to component being rendered
+    # block: optional block to be called within the view context
+    #
+    # returns HTML that has been escaped with the ERB pipeline
+    #
+    # Example subclass:
+    #
+    # app/components/my_component.rb:
+    # class MyComponent < ActionView::Component
+    #   def initialize(title:)
+    #     @title = title
+    #   end
+    # end
+    #
+    # app/components/my_component.html.erb
+    # <span title="<%= @title %>">Hello, <%= content %>!</span>
+    #
+    # In use:
+    # <%= render MyComponent.new(title: "greeting") do %>world<% end %>
+    # returns:
+    # <span title="greeting">Hello, world!</span>
+    #
+    def render_in(view_context, *args, &block)
       self.class.compile
       @content = view_context.capture(&block) if block_given?
       validate!
-      rendered_template
+      call
     end
 
+    def initialize(*); end
+
     def self.compile
       @compiled ||= nil
       return if @compiled
 
       class_eval(
-        "def rendered_template; @output_buffer = ActionView::OutputBuffer.new; " +
+        "def call; @output_buffer = ActionView::OutputBuffer.new; " +
         ActionView::Template::Handlers::ERB.erb_implementation.new(template, trim: true).src +
         "; end"
       )
 
       @compiled = true
     end
 
-  private
+    def self.template
+      filename = self.instance_method(:initialize).source_location[0]
+
+      raise NotImplementedError.new("Subclasses of ActionView::Component must implement #initialize") if filename == __FILE__
+
+      filename_without_extension = filename[0..-(File.extname(filename).length + 1)]
+
+      erb_template_path = filename_without_extension+".html.erb"
```

Comment:
@joelhawksley I see the support for `.html.erb` is hardwired in. Is there a way we could rely on standard Rails template rendering that supports other template languages (such as slim)?

A nice reference can be found in [Dissasembing Rails – Template Rendering](https://medium.com/rubyinside/disassembling-rails-template-rendering-1-51795f579577).

Would that be a next step? Would you be interested in PR that tries to use standard Rails-y template rendering?

Reply:
@tomasc you're 100% right here, this is totally a hack 😄 

Sure, send a PR, as long as you are OK being marked as a co-author if your work is incorporated into any upstream commits to Rails 😄

Reply:
@joelhawksley I started something here https://github.com/joelhawksley/actionview-component-demo/pull/12. However, I could use some help with properly evaluating nested components in the context of the root component. See the PR for more details.